avatars/processors/inter_record_bounded_cumulated_difference.py

```diff
@@ -0,0 +1,260 @@
+    Examples
+    --------
+    >>> df = pd.DataFrame({
+    ...    "id": [1, 2, 1, 1, 2, 2],
+    ...    "value": [1025, 20042, 1000, 1130, 20000, 20040],
+    ... })
+    >>> processor = InterRecordBoundedCumulatedDifferenceProcessor(
+    ...    id_variable='id',
+    ...    target_variable='value',
+    ...    new_first_variable_name='first_value',
+    ...    new_difference_variable_name='difference_to_bound',
+    ...    should_round_output=True
+    ...    )
+    >>> processor.preprocess(df)
+       id  first_value  difference_to_bound
+    0   1         1025             0.000000
+    1   2        20042             0.000000
+    2   1         1025            -1.000000
+    3   1         1025             0.006827
+    4   2        20042            -0.002206
+    5   2        20042             0.952381
+
+    The postprocess allows you to transform some preprocessed data back into its original format
+
+    >>> preprocessed_df = pd.DataFrame({
+    ...    "id": [1, 2, 1, 1, 2, 2],
+    ...    "first_value": [1025, 20042, 1025, 1025, 20042, 20042],
+    ...    "difference_to_bound": [0, 0, -1, 0.006827, -0.002206, 0.952381],
+    ... })
+    >>> processor.postprocess(df, preprocessed_df)
+       id  value
+    0   1   1025
+    1   2  20042
+    2   1   1000
+    3   1   1130
+    4   2  20000
+    5   2  20040
+    """
+
+    def __init__(
+        self,
+        id_variable: str,
+        target_variable: str,
+        new_first_variable_name: str,
+        new_difference_variable_name: str,
+        should_round_output: bool = False,
+    ):
+        self.id_variable = id_variable
+        self.target_variable = target_variable
+        self.new_first_variable_name = new_first_variable_name
+        self.new_difference_variable_name = new_difference_variable_name
+        self.round_output = should_round_output
+
+    def preprocess(self, df: pd.DataFrame) -> pd.DataFrame:
+        if self.id_variable not in df.columns.values:
+            raise ValueError(
+                f"Expected a valid `id_variable`, got {self.id_variable} instead"
+            )
+
+        if self.target_variable not in df.columns.values:
+            raise ValueError(
+                f"Expected a valid `target_variable`, got {self.target_variable} instead"
+            )
+
+        if df[self.id_variable].isnull().values.any():
+            raise ValueError(
+                "Expected no missing values for id variable, got column with nulls instead"
+            )
+
+        if df[self.target_variable].isnull().values.any():
+            raise ValueError(
+                "Expected no missing values for target variable, got column with nulls instead"
+            )
+
+        working = df.copy()
+
+        # determine lb and ub
+        working["lb"] = min(working[self.target_variable])
+        working["ub"] = max(working[self.target_variable])
+
+        # store first value of target variable
```

Comment:
we might lack a unit test to check if the first value per id number is working as we expect. Don't know if it's worth making a function out of it though.

Reply:
added an assert to test that preprocess is as expected (and so contains the expected `first_value`